File: Go/chattest/chat_test.go
package chattest

import (
	"fmt"
	"math/rand"
	"os"
	"testing"
	"time"
)

//execution order for a test
func TestMain(m *testing.M) {
	mySetupFunction()
	retCode := m.Run()
	myTeardownFunction()
	os.Exit(retCode)
}

//start server before test
func mySetupFunction() {
	go startServer()
}

//nothing to do after test
func myTeardownFunction() {

}

//utitlity func to assert actual vs expected for a test
func assertEquals(t *testing.T, actual, expected int) {
	t.Helper()
	if actual != expected {
		t.Errorf("actual '%d' expected '%d'", actual, expected)
	}
}

//Two clients are created. First client sends a message to the server which then broadcasts to the second client
//Both clients exit after this and the broadcast latency is recorded
func Test1ServerBroadcastLatency(t *testing.T) {

	start := time.Now()
	data := make([]string, 2)
	data[0] = "[bye]"//single datapoint that will be used
	done := make(chan bool, 2)
	//start both the clients
	go startClient(data, done)
	go startClient(nil, done)
	count := 0
	//wait for the clients to exit
	for i := 1; i <= 2; i++ {
		<-done
		count = count + 1
	}
	assertEquals(t, count, 2)
	elapsed := time.Since(start)
	//elapsed time depicts the server broadcast latency
	fmt.Println(elapsed)

}

/* 'numclients(=100)' clients are created. Each clients sends 'datapoints(=2)' messages to the server which then broadcasts the msg to 
the remaining clients. All the clients exit after this process and we measure the time recorded to be an indicator 
of the chatroom application's bandwidth */
func Test2ClientChatUsingMultipleClients(t *testing.T) {

	//code to create random string for the datapoint
	const charset = "abcdefghijklmnopqrstuvwxyz" +
		"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

	const numclients = 100
	const datapoints = 2

	var seededRand = rand.New(
		rand.NewSource(time.Now().UnixNano()))

	stringmaker := func(length int, charset string) string {
		b := make([]byte, length)
		for i := range b {
			b[i] = charset[seededRand.Intn(len(charset))]
		}
		return string(b)
	}
	getstring := func(length int) string {
		return stringmaker(length, charset)
	}

	start := time.Now()
	data := make([]string, datapoints)
	done := make(chan bool, numclients)
	//start all the clients and pass the two msgs to each of them to broadcast
	for i := 0; i < numclients; i++ {
		for j := 0; j < datapoints; j++ {
			data[j] = getstring(10)
		}
		go startClient(data, done)
	}
	//wait for all the clients to finish
	count := 0
	for i := 1; i <= numclients; i++ {
		<-done
		count = count + 1
	}
	assertEquals(t, count, numclients)
	elapsed := time.Since(start)
	//measure the elapsed time
	fmt.Println(elapsed)
}

File: Java/MultiThreaded Chat/multichat/ChatTest.java
package multichat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.Assert;
import org.junit.Before;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runners.MethodSorters;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class ChatTest {
	
	private final String hostName = "127.0.0.1";
	private final static int port = 59001;
	private final int num_clients = 100;
	private final static ChatServer chatServer = new ChatServer(port);
	//keeps track of writer threads in order to know when the client quit
	public static volatile List<WriteThread> writers = new ArrayList<>();
	//for the clients to write to when they quit
	public static AtomicInteger clientQuitCount = new AtomicInteger(0);
	
	//start server before tests
	static {
		new Thread(() -> chatServer.execute()).start();
		try {
			Thread.sleep(2000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
	
	//calls before each test. Resets client and writer data.
	@Before
	public void reset() {
		clientQuitCount = new AtomicInteger(0);
		writers.clear();
	}
	
	//Two clients are created. First client sends a message to the server which then broadcasts to the second client
	//Both clients exit after this and the broadcast latency is recorded
	@Test
	public void _1_testServerBroadcastLatency() {
	
		Long start = System.currentTimeMillis();
		ChatClient client1 = new ChatClient(hostName, port), client2 = new ChatClient(hostName, port);
		client1.setTestData(new ArrayList<>(Arrays.asList("[bye]")));
		client1.setUserName("1");
		client2.setUserName("2");
		//start clients 1 and 2
		client1.execute();
		client2.execute();
		//wait for them to finish
		for(WriteThread wt : writers) {
			try {
				wt.join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}	
		Assert.assertEquals(1, clientQuitCount.get());
		//elapsed time depicts the broadcast latency of the server
		System.out.println("total elapsed time ====================================  " + (System.currentTimeMillis()-start) + " ms");
	}
	
	/* 'num_clients(=100)' clients are created. Each clients sends 2 messages to the server which then broadcasts the msg to 
	    the remaining clients. All the clients exit after this process and we measure the time recorded to be an indicator 
	    of the chatroom application's bandwidth */
	@Test
	public void _2_testClientChatUsingMultipleClients() {
		
		Long start = System.currentTimeMillis();
		List<ChatClient> chatClients = new ArrayList<>();
		for(int i=0;i<num_clients; i++)
			chatClients.add(new ChatClient(hostName, port));
		
		//start 'num_clients' clients and send message to the server
		for(int i=0;i<num_clients; i++) {
			ChatClient client = chatClients.get(i);
			client.setUserName(Integer.toString(i));
			List<String> testmsg = new ArrayList<>();
			testmsg.add(i + " test message");
			testmsg.add("[bye]");
			client.setTestData(testmsg);
			client.execute();
		}
		//wait for the clients to finish
		for(WriteThread wt : writers) {
			try {
				wt.join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		Assert.assertEquals(num_clients, clientQuitCount.get());
		//measure the elapsed time
		System.out.println("total elapsed time ==================================== " + (System.currentTimeMillis()-start) + " ms");
	}
}
